test/system/505-networking-pasta.bats

```diff
@@ -209,34 +209,46 @@ function pasta_test_do() {
         local expect="$(for i in $(seq ${seq}); do printf "x"; done)"
     fi
 
-    # Set retry/initial delay for client to connect
-    local delay=1
-    if [ ${ip_ver} -eq 6 ] && [ "${addr}" != "::1" ]; then
-        # Duplicate Address Detection on link-local
-        delay=3
-    elif [ "${proto}" = "udp" ]; then
-        # With Podman up, and socat still starting, UDP clients send to nowhere
-        delay=2
-    fi
+    # start server
+    local cname="c-socat-$(safename)"
+    run_podman run -d --name="$cname" --net=pasta"${pasta_spec}" -p "${podman_spec}" "${IMAGE}" \
+                   sh -c 'for port in $(seq '"${xseq}"'); do '\
+'                             socat -u '"${bind}"' '"${recv}"' & '\
+'                         done; wait'
 
-    # Now actually run the test: client,
-    for one_port in $(seq ${seq}); do
-        local connect="${proto_upper}${ip_ver}:[${addr}]:${one_port}"
-        [ "${proto}" = "udp" ] && connect="${connect},shut-null"
+    run_podman container inspect --format "{{.NetworkSettings.SandboxKey}}" $cname
+    netns="$output"
+
+    # Make sure all ports in the contianer are bound
+    for cport in $(seq ${xseq}); do
+        retries=50
+        while [[ $retries -gt 0 ]]; do
+            run_podman unshare nsenter -n"$netns" ss -Htuln -$ip_ver sport = $cport
+            if [[ $output =~ $cport ]]; then
```

Comment:
For some reason, this isn't working on IPv6 "tap" tests. It works for IPv4, and for IPv6 loopback tests. I'm not really familiar with non-POSIX shell syntax, so here comes the relevant output of one failing test as example:

```
 ✗ [505] Single UDP port forwarding, IPv6, tap
   (from function `run_podman' in file helpers.bash, line 527,
    from function `pasta_test_do' in file 505-networking-pasta.bats, line 247,
    in test file 505-networking-pasta.bats, line 634)
     `pasta_test_do' failed
   
   [16:59:47.268721829] $ podman info --format {{.Host.Pasta.Executable}}
   [16:59:48.118448097] /usr/bin/pasta
   
   [16:59:48.179590796] $ podman run -d --name=c-socat-t13-1bowvp8k --net=pasta -p [2a01:4f8:222:904::2]:5095:5095/udp quay.io/libpod/testimage:20240123 sh -c for port in $(seq 5095 5095); do                              socat -u UDP6-LISTEN:${port},null-eof STDOUT &                          done; wait
   [16:59:48.291610727] d07505ed44b255ca84dff14d3fd0f7c527fb2c93bbcafc1f8aebb320686cb9e8
   
   [16:59:48.296102453] $ podman container inspect --format {{.NetworkSettings.SandboxKey}} c-socat-t13-1bowvp8k
   [16:59:48.325410606] /run/user/1000/netns/netns-760759d3-9a23-882f-bb67-63c86c556a40
   
   [16:59:48.332287966] $ podman unshare nsenter -n/run/user/1000/netns/netns-760759d3-9a23-882f-bb67-63c86c556a40 ss -Htuln -6 sport = 5095
   [16:59:48.398745612] udp UNCONN 0      0      *:5095 *:*
   
   [16:59:48.411588810] $ podman logs --follow c-socat-t13-1bowvp8k
   [17:01:48.439216198] timeout: sending signal TERM to command ‘podman’
   [17:01:48.442469452] [ rc=124 (** EXPECTED 0 **) ]
   *** TIMED OUT ***
   # [teardown]
```

perhaps you could simply check if the output is non-empty `[ -n "${output}" ] && break` instead?

Reply:
I think the issue is that I didn't quote the vars

Reply:
`[ -n "${output}" ] && break` doesn't work either, actually :(

Reply:
I cannot test this on my v6 setup it seems the test still require a public routeable address even though pasta doesn't anymore.

Reply:
Perhaps we could drop that requirement, but we would still need to check for some form of IPv6 support.

Anyway, let me debug this on my setup.

Reply:
added `sudo ip -6 route add  blackhole default` which is enough to convice the test that I have v6, some other test is failing because of that but it does work for the port forward tests and I don't see any failures there

The assert is definitely working as I set some incorrect port to test. The timeout means your container didn't exit which means it didn't get the right traffic. I do not know why that would be.

Reply:
Ah, that could be because the port test succeeded too early, I guess.

Reply:
Nope, it's because even if we set up IPv6 addresses with `nodad`:

```
$ pasta --config-net -a 2001:db8::1 -I x -- ip -6 addr show x
2: x: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 65520 qdisc fq_codel state UNKNOWN group default qlen 1000
    inet6 2001:db8::1/64 scope global nodad 
       valid_lft forever preferred_lft forever
    inet6 fe80::c0:e5ff:fede:7eaa/64 scope link tentative proto kernel_ll 
       valid_lft forever preferred_lft forever
```

we still need to wait a bit before it's usable (a `sleep 2` after the port test "fixes" this). I'm trying to find out why.

Reply:
This works:

```
$ (sleep 0.1; echo x | socat -u STDIN TCP6:localhost:10122; ) & pasta --config-net --no-map-gw -t 10122 -u none -T none -U none -- socat -u TCP6-LISTEN:10122 STDOUT
[2] 771470
x
[2]-  Done                    ( sleep 0.1; echo x | socat -u STDIN TCP6:localhost:10122 )
```

so it only seems to happen with Podman for some obscure reason...

Reply:
with:
```
--- a/test/system/505-networking-pasta.bats
+++ b/test/system/505-networking-pasta.bats
@@ -195,7 +195,7 @@ function pasta_test_do() {
         local podman_spec="[${addr}]:${port}:${xport}/${proto}"
         ;;
     *)
-        local pasta_spec=
+        local pasta_spec=":-p,/tmp/podman.pcap"
         local podman_spec="[${addr}]:${port}:${xport}/${proto}"
         ;;
     esac
```

```
$ tshark -r /tmp/podman.pcap 
    1   0.000000           :: → ff02::16     ICMPv6 110 Multicast Listener Report Message v2
    2   0.055984           :: → ff02::16     ICMPv6 110 Multicast Listener Report Message v2
    3   0.103942      fe80::1 → fe80::aaa1:59ff:fe8e:d7b6 TCP 82 40384 → 5565 [SYN] Seq=0 Win=65535 Len=0 MSS=61440 WS=256
    4   0.363974           :: → ff02::1:ff45:e16e ICMPv6 86 Neighbor Solicitation for fe80::188d:54ff:fe45:e16e
    5   1.387994 fe80::188d:54ff:fe45:e16e → ff02::16     ICMPv6 110 Multicast Listener Report Message v2
    6   1.388012 fe80::188d:54ff:fe45:e16e → ff02::2      ICMPv6 70 Router Solicitation from 1a:8d:54:45:e1:6e
    7   2.187984 fe80::188d:54ff:fe45:e16e → ff02::16     ICMPv6 110 Multicast Listener Report Message v2
```

but if I add the `sleep 2` back:

```
$ tshark -r /tmp/podman.pcap 
    1   0.000000           :: → ff02::16     ICMPv6 110 Multicast Listener Report Message v2
    2   0.672237           :: → ff02::1:ffd0:b6f0 ICMPv6 86 Neighbor Solicitation for fe80::e1:42ff:fed0:b6f0
    3   0.704217           :: → ff02::16     ICMPv6 110 Multicast Listener Report Message v2
    4   1.696012 fe80::e1:42ff:fed0:b6f0 → ff02::16     ICMPv6 110 Multicast Listener Report Message v2
    5   1.696030 fe80::e1:42ff:fed0:b6f0 → ff02::2      ICMPv6 70 Router Solicitation from 02:e1:42:d0:b6:f0
    6   2.100413      fe80::1 → fe80::e1:42ff:fed0:b6f0 TCP 82 38212 → 5428 [SYN] Seq=0 Win=65535 Len=0 MSS=61440 WS=256
    7   2.100434 fe80::e1:42ff:fed0:b6f0 → ff02::1:ff00:1 ICMPv6 86 Neighbor Solicitation for fe80::1 from 02:e1:42:d0:b6:f0
    8   2.100457      fe80::1 → fe80::e1:42ff:fed0:b6f0 ICMPv6 86 Neighbor Advertisement fe80::1 (rtr, sol, ovr) is at a8:a1:59:8e:d7:b6
    9   2.100465 fe80::e1:42ff:fed0:b6f0 → fe80::1      TCP 82 5428 → 38212 [SYN, ACK] Seq=0 Ack=1 Win=65460 Len=0 MSS=65460 WS=4096
```

...for some reason, the target address of the neighbour solicitation is not correct (it should be `fe80::1`) without the delay. It looks like the default gateway is not set up on time or something like that...

Reply:
Do we even need DAD given we use --config-net it should be safe to assume there are no conflicts? I know we disabled DAD for this reason in other parts because it just to slow to make networking directly after start work.

Reply:
Ah, no, the source address is missing (`::`), and we [silently ignore](https://passt.top/passt/tree/ndp.c?id=baccfb95ce0e30f64d052f710cd5fedc8c7426c1#n70) the neighbour solicitation in that case, according to [RFC 4862, 5.4.3](https://datatracker.ietf.org/doc/html/rfc4862#section-5.4.3), as that indicates a probe for Duplicate Address Detection.

I'm afraid this will need more involved debugging and possibly a fix or a workaround (...why is the kernel sending that solicitation so late that it comes after a `SYN` segment?). If you need to merge this quickly, I would suggest keeping a `sleep 2` for IPv6 *after* the bound port test, that's robust:

```
         # Ports are bound: we can start the client in the background
+       [ ${ip_ver} = 6 ] && sleep 2
```

Reply:
> Do we even need DAD given we use --config-net it should be safe to assume there are no conflicts? I know we disabled DAD for this reason in other parts because it just to slow to make networking directly after start work.

No, we don't need it because the address *is* duplicated, by default, so we [set](https://passt.top/passt/tree/netlink.c?id=baccfb95ce0e30f64d052f710cd5fedc8c7426c1#n783) `IFA_NO_DAD` on the interface.

That works, in general, but if we send a `SYN` segment "too early", we seem to trigger DAD somehow.

Reply:
ack thanks, I will come back to this on Monday. I guess paying the price of sleep 2 only for v6 is already a good improvement so I would be fine with that for now.

Reply:
> > Do we even need DAD given we use --config-net it should be safe to assume there are no conflicts? I know we disabled DAD for this reason in other parts because it just to slow to make networking directly after start work.
> 
> No, we don't need it because the address _is_ duplicated, by default, so we [set](https://passt.top/passt/tree/netlink.c?id=baccfb95ce0e30f64d052f710cd5fedc8c7426c1#n783) `IFA_F_NODAD` on the interface.

Sorry, wrong pointer, by default we do this: https://passt.top/passt/tree/netlink.c?id=baccfb95ce0e30f64d052f710cd5fedc8c7426c1#n851 (set `IFA_F_NODAD` on all the addresses we copy from the host).

However, we don't copy link-local addresses (of course), and here we actually see DAD performed on a link-local address (which is also what we're using for the test). Instead of ::1, I guess we should use be using a global unicast address (as destination) to reach the container in these tests. I'll play with that.

Reply:
Nope, sorry, even if we use global unicast addresses as destination, currently pasta translates them to link-local addresses. Maybe we can avoid to do that if we're sure global addresses would work (in Podman's case, we just configured them...), but I have to look into it. For the moment, this is the best I can offer:

```
diff --git a/test/system/505-networking-pasta.bats b/test/system/505-networking-pasta.bats
index c6bccc1b7..21abe795b 100644
--- a/test/system/505-networking-pasta.bats
+++ b/test/system/505-networking-pasta.bats
@@ -230,6 +230,15 @@ function pasta_test_do() {
         assert $retries -gt 0 "Timed out waiting for bound port $cport in container"
     done
 
+    if [ ${ip_ver} -eq 6 ] && [ ${iftype} = "tap" ]; then
+        # For IPv6 tests via tap interface, we use pairs of link-local
+        # addresses to communicate. While we disable DAD on all the
+        # (global unicast) addresses we copy from the host, or
+        # otherwise set, link-local addresses are automatically added
+        # by the kernel with DAD, so we need to wait until it's done.
+        sleep 2
+    fi
+
     for hport in $(seq ${seq}); do
         local connect="${proto_upper}${ip_ver}:[${addr}]:${hport}"
         [ "${proto}" = "udp" ] && connect="${connect},shut-null"
```

Reply:
thanks, added

Reply:
We chatted with @dgibson about this... the main reason why we force a link-local address as destination when mapping addresses from an IPv6 local connection is that the source is a link-local address, and he observed issues when mixing link-local (as source) and global (as destination) addresses, which was the previous behaviour (not clean, but generally working/not forbidden anyway).

It sounds overall a good idea to maintain this behaviour. I will try one of these three things: 1. modifying netlink attributes of the link-local address as we bring up the interface to set `IFA_F_NODAD`, 2. enable optimistic duplicate address detection (via procfs) in the container before we configure the interface, 3. set up link-local addresses in pasta itself. If something like that works, we'll then be able to drop the additional delay in this case.

Reply:
ok so that means we still need to work around here until it gets fixed on your side

Reply:
Well, it's not a workaround and it wouldn't be a fix :) ...but yes.

Reply:
@Luap99 with Fedora's `passt-0^20240821.g1d6142f-1.fc40`, `passt-0^20240821.g1d6142f-1.fc41`, `passt-0^20240821.g1d6142f-1.fc42`, and Debian's `passt-0.0~git20240814.61c0b0d-1`, we can drop the additional delay here. I can also submit a pull request once the CI images are updated.

Reply:
As of [yesterday](https://github.com/containers/automation_images/pull/381#issuecomment-2299387093) pasta was 2024-08-14 on debian, f40, and rawhide; but still 06-24 on f39. I've just +1'ed its bodhi.